File: src/main/java/cn/xunyard/idea/util/ObjectUtils.java
package cn.xunyard.idea.util;

import java.util.function.Function;
import java.util.function.Supplier;

/**
 * @author <a herf="mailto:wuqi@terminus.io">xunyard</a>
 * @date 2019-12-22
 */
public class ObjectUtils {

    public static <T> T firstNonNull(T first, T other) {
        if (first != null) {
            return first;
        }

        if (other == null) {
            throw new IllegalArgumentException("both.argument.is.null");
        }

        return other;
    }

    public static <T> T firstNonNull(T first, Supplier<T> supplier) {
        if (first != null) {
            return first;
        }

        if (supplier == null) {
            throw new IllegalArgumentException("supplier.is.null");
        }

        return supplier.get();
    }

    public static <T, R> R takeIfNonNull(T source, Function<T, R> take) {
        if (source == null) {
            return null;
        }

        return take.apply(source);
    }

    public static boolean takeBoolean(Object obj) {
        return obj != null && Boolean.parseBoolean(obj.toString());
    }

    public static String smoothStr(String s) {
        if (s == null) {
            return null;
        }

        String[] splits = s.split("\n");
        StringBuilder sb = new StringBuilder();
        for (String split : splits) {
            sb.append(split);
        }
        return sb.toString();
    }
}

File: src/main/kotlin/cn/xunyard/idea/coding/util/ObjectUtils.kt
package cn.xunyard.idea.coding.util

import cn.xunyard.idea.coding.util.AssertUtils.isEmpty
import cn.xunyard.idea.coding.util.AssertUtils
import java.lang.StringBuilder
import java.util.function.Function
import java.util.function.Supplier

/**
 * @author <a herf="mailto:wuqi@terminus.io">xunyard</a>
 * @date 2019-12-22
 */
object ObjectUtils {
    fun <T> firstNonNull(first: T?, other: T?): T {
        if (first != null) {
            return first
        }
        requireNotNull(other) { "both.argument.is.null" }
        return other
    }

    fun <T> firstNonNull(first: T?, supplier: Supplier<T>?): T {
        if (first != null) {
            return first
        }
        requireNotNull(supplier) { "supplier.is.null" }
        return supplier.get()
    }

    fun <T, R> takeIfNonNull(source: T?, take: Function<T, R>): R? {
        return if (source == null) {
            null
        } else take.apply(source)
    }

    fun takeBoolean(obj: Any?): Boolean {
        return obj != null && java.lang.Boolean.parseBoolean(obj.toString())
    }

    fun smoothStr(s: String?): String? {
        if (s == null) {
            return null
        }
        val splits = removeQuotation(s).split("\n".toRegex()).toTypedArray()
        val sb = StringBuilder()
        for (split in splits) {
            sb.append(split)
        }
        return sb.toString()
    }

    fun removeQuotation(str: String): String {
        if (isEmpty(str)) {
            return str
        }
        val trim = str.trim { it <= ' ' }
        return if (trim.startsWith("\"")) {
            trim.substring(1, str.length - 1)
        } else trim
    }
}